compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParser.java

```diff
@@ -820,19 +820,16 @@ private STNode parseTopLevelNode(STNode metadata, STNode publicQualifier, List<S
                 // Here we assume that after recovering, we'll never reach here.
                 // Otherwise the tokenOffset will not be 1.
                 if (isModuleVarDeclStart(1)) {
-                    reportInvalidQualifier(publicQualifier);
-                    return parseModuleVarDecl(metadata, qualifiers);
+                    return parseModuleVarDecl(metadata, publicQualifier, qualifiers);
                 }
                 // fall through
             default:
-                if (isPossibleServiceDecl(qualifiers)) {
-                    reportInvalidQualifier(publicQualifier);
+                if (publicQualifier == null && isPossibleServiceDecl(qualifiers)) {
                     return parseServiceDeclOrVarDecl(metadata, qualifiers);
```

Comment:
I think we have to do the `public` qualifier validation inside the `parseServiceDeclOrVarDecl()`. As the method name implies, it could either be a `service-decl` or `var-decl`. 
i.e. `service-decl` with `public` qualifier will go on the wrong path.